### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by daniel-vega-86**
## Summary

[Change!] Implement the /users endpoint, set validations and hash password. 
Hashing the password cost me a lot of time for the race condition error in the way i usually hash (leave the code commented) and i have to try a differente way. Also i still haven't figured out how the errors logger works, so i think this is still missing.

## Known Issues

[Change!] None.

## Trello Card

[Change!] https://trello.com/c/FyBXqvci/35-sign-up

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const usersService = require('../services/users');
+
+const signUp = async ({ body }, res) => {
+  try {
+    const user = await usersService.createUser(body);
+    console.info(`username: ${user.email}`);
+    res.status(201).send(user);
```
Comment (dbenavidesv) [inline app/controllers/users.js]: It's always a good practice to add a return here `return res.status(201).send(user);`. This is because it helps to prevent flow issues with the express router, if you'd like me to explain better, dm me :D

File: app/controllers/users.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const usersService = require('../services/users');
+
+const signUp = async ({ body }, res) => {
```
Comment (dbenavidesv) [inline app/controllers/users.js]: This is more of an opinion but, although it's totally fine to deconstruct here, I prefer not to do it, to make clear that this function is a controller and besides there are some properties in the res object that you may need in some cases

File: app/controllers/users.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  try {
+    const user = await usersService.createUser(body);
+    console.info(`username: ${user.email}`);
+    res.status(201).send(user);
+  } catch (e) {
+    res.status(400).send(e.message);
```
Comment (dbenavidesv) [inline app/controllers/users.js]: Again, this is fine, but with express middlewares you can add error handling, this base project has it (`app/middlewares/errors.js`). To use it in any controller you would have to make use of the `next()` express function, this function detects if what you send to it is an error and routes it to the error handling middleware. Do that here :D

File: app/controllers/weets.js [hunk 4]
```diff
@@ -1,8 +1,8 @@
-const service = require('../services/weets');
+const weetsService = require('../services/weets');
 
 const generateWeet = async (_, res) => {
   try {
-    const weet = await service.fetchWeet();
+    const weet = await weetsService.fetchWeet();
     res.status(200).send(weet);
   } catch (e) {
     res.status(400).send(e.message);
```
Comment (dbenavidesv) [inline app/controllers/weets.js]: Now apply the previous comments here as well!

File: app/helpers/hashPassword.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+const bcrypt = require('bcrypt');
+
+const hashPassword = async password => {
```
Comment (dbenavidesv) [inline app/helpers/hashPassword.js]: A good way of thinking about helper functions is that they should be as generic as posible, to make them more reusable. So instead of making a `hashPassword` function it's preferable to have a `hashSomething` or `hashString` function. What do you think?
P.D.: The same goes to the file name

File: app/middlewares/validations.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+const { checkSchema, validationResult } = require('express-validator');
+
+const verifySchema = schema => checkSchema(schema);
+
+const validateResult = (req, res, next) => {
+  const err = validationResult(req);
+  if (err.isEmpty()) {
+    next();
+  } else {
+    res.status(409).send(err.mapped());
```
Comment (dbenavidesv) [inline app/middlewares/validations.js]: Use `next()` here. And at this point, look how to create a custom error for this type of exception. Also `409` is not the correct code here, use this for reference [HTTP CODES](https://www.restapitutorial.com/httpstatuscodes.html)

File: app/schemas/user_signup.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  },
+  lastName: {
+    exists: true,
+    notEmpty: true,
+    errorMessage: 'Last name cannot be empty!'
+  },
+  email: {
+    exists: true,
+    isEmail: { errorMessage: 'Not a valid email.' },
+    contains: { options: '@wolox.co', errorMessage: 'Email does not belong wolox domain.' },
```
Comment (dbenavidesv) [inline app/schemas/user_signup.js]: There are several wolox domains: wolox.(co|ar|cl|mx), find a way to accept all of these.

File: app/schemas/user_signup.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    notEmpty: { errorMessage: 'email cannot be empty!' },
+    custom: {
+      options: async value => {
+        const user = await db.User.findOne({ where: { email: value } });
+        if (user) {
+          throw new Error('Email is already registered.');
```
Comment (dbenavidesv) [inline app/schemas/user_signup.js]: This validator is nice, but I think throwing this error like this would make it have an 500 http code, which is not very telling. Better to throw a custom error here :)

File: app/controllers/users.js (hunk 1 above)
Comment (daniel-vega-86) [reply app/controllers/users.js]: Try to use it but eslint throw an error.

File: app/schemas/user_signup.js (hunk 8 above)
Comment (daniel-vega-86) [reply app/schemas/user_signup.js]: this wil throw a validation error with the code stablish for that, previosly 409 but i'm going to change it. If a different code must be return for this particular case tell me and i will make it. :)

Comment (dbenavidesv) [reply app/schemas/user_signup.js]: The thing is, `409` is the right code when the email already exists but not for when an input is incorrect, that would be `422`, but throwing inside this validator I think will always be wrapped in a bad input error

File: app/controllers/users.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+const usersService = require('../services/users');
+const logger = require('../logger');
+
+const signUp = async (req, res, next) => {
+  try {
+    const user = await usersService.createUser(req.body);
+    logger.info(`username: ${user.email}`);
+    res.status(201).send(user);
+  } catch {
+    next();
```
Comment (dbenavidesv) [inline app/controllers/users.js]: This `next()` should have the error that is catched as an argument, should be something like this:
```
catch (error) {
  next(error);
}

File: app/controllers/weets.js [hunk 10]
```diff
@@ -1,11 +1,11 @@
...
-const generateWeet = async (_, res) => {
+const generateWeet = async (_, res, next) => {
   try {
-    const weet = await service.fetchWeet();
+    const weet = await weetsService.fetchWeet();
     res.status(200).send(weet);
-  } catch (e) {
-    res.status(400).send(e.message);
+  } catch {
+    next();
```
Comment (dbenavidesv) [inline app/controllers/weets.js]: Same here

File: app/helpers/hashString.js [hunk 11]
```diff
@@ -0,0 +1,8 @@
+const bcrypt = require('bcrypt');
+
+const hashString = async password => {
```
Comment (dbenavidesv) [inline app/helpers/hashString.js]: I think you forgot to rename this parameter

File: app/middlewares/errors.js [hunk 12]
```diff
@@ -5,7 +5,8 @@ const DEFAULT_STATUS_CODE = 500;
 
 const statusCodes = {
   [errors.DATABASE_ERROR]: 503,
-  [errors.DEFAULT_ERROR]: 500
+  [errors.DEFAULT_ERROR]: 500,
+  [errors.SCHEMA_VALIDATION]: 422
```
Comment (dbenavidesv) [inline app/middlewares/errors.js]: Nice

File: app/errors.js [hunk 13]
```diff
@@ -8,3 +8,6 @@ exports.databaseError = message => internalError(message, exports.DATABASE_ERROR
 
 exports.DEFAULT_ERROR = 'default_error';
 exports.defaultError = message => internalError(message, exports.DEFAULT_ERROR);
+
+exports.SCHEMA_VALIDATION = 'schema_validation';
+exports.schemaValiation = errors => internalError(errors, exports.SCHEMA_VALIDATION);
```
Comment (dbenavidesv) [inline app/errors.js]: I know it's a bit wordy, but generally error functions should have de keyword `error` in their names, it makes the code easier to read. It should be `schemaValidationError`, the same goes for `SCHEMA_VALIDATION_ERROR`

File: app/middlewares/validations.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+const { checkSchema, validationResult } = require('express-validator');
+const errors = require('../errors');
+
+const verifySchema = schema => checkSchema(schema);
+
+const validateResult = (req, _, next) => {
+  const err = validationResult(req);
+  if (!err.isEmpty()) {
+    throw errors.schemaValiation(err.mapped());
```
Comment (dbenavidesv) [inline app/middlewares/validations.js]: Here you should use `return next(error.schema....)`, In middleware/controller functions `next()` should be used instead of throw

File: app/middlewares/uniqueEmail.js [hunk 15]
```diff
@@ -0,0 +1,12 @@
+const db = require('../models');
+const errors = require('../errors');
+
+const uniqueEmail = async (req, res, next) => {
```
Comment (dbenavidesv) [inline app/middlewares/uniqueEmail.js]: Generally functions should be actions, so the code is easier to read, maybe `checkUniqueEmail` or `validateEmail` or `checkIfEmailExists`, And maybe use a more generic name for the file.

File: app/schemas/user_signup.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  lastName: {
+    exists: true,
+    notEmpty: true,
+    errorMessage: 'Last name cannot be empty!'
+  },
+  email: {
+    exists: true,
+    isEmail: { errorMessage: 'Not a valid email.' },
+    matches: {
+      options: [/^[\w-.]+@wolox(.)+[(co|ar|cl|mx)]{2,2}$/],
```
Comment (dbenavidesv) [inline app/schemas/user_signup.js:16]: nice

File: app/services/users.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
+const db = require('../models');
+
+const createUser = body =>
+  new Promise((resolve, reject) => {
```
Comment (dbenavidesv) [inline app/services/users.js]: You don't need to wrap this in a promise since the `User.newUser()` function already returns a promise

File: app/services/users.js [hunk 18]
```diff
@@ -0,0 +1,8 @@
+const db = require('../models');
+
+const createUser = body =>
+  db.User.newUser(body)
+    .then(user => user)
+    .catch(error => error);
```
Comment (dbenavidesv) [inline app/services/users.js]: Here use `Promise.reject()` and throw a custom error for database errors

File: app/services/weets.js [hunk 19]
```diff
@@ -4,10 +4,7 @@ const fetchWeet = () =>
   api
     .get('/random')
     .then(({ data }) => data)
-    .catch(error => {
-      console.error(error);
-      return Promise.reject(error);
-    });
+    .catch(error => Promise.reject(error));
```
Comment (dbenavidesv) [inline app/services/weets.js]: And here a custom error as well, but for external APIs

**Review by dbenavidesv — APPROVED** [review]
